src/Microsoft.Azure.WebJobs.Host/Hosting/DrainModeManager.cs

```diff
@@ -0,0 +1,94 @@
+{
+    internal class DrainModeManager : IDrainModeManager
+    {
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly ILogger _logger;
+        private ConcurrentDictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
+
+        public DrainModeManager(ILoggerFactory loggerFactory = null)
+        {
+            _loggerFactory = loggerFactory;
+            _logger = _loggerFactory.CreateLogger<DrainModeManager>();
+            _cancellationTokenSources = new ConcurrentDictionary<Guid, CancellationTokenSource>();
+        }
+
+        public bool IsDrainModeEnabled { get; private set; } = false;
+
+        public ICollection<IListener> Listeners { get; private set; } = new Collection<IListener>();
+
+        public void RegisterListener(IListener listener)
+        {
+            Listeners.Add(listener);
+        }
+
+        public void RegisterTokenSource(Guid guid, CancellationTokenSource tokenSource)
+        {
+            _cancellationTokenSources.TryAdd(guid, tokenSource);
+        }
+
+        public void UnRegisterTokenSource(Guid guid)
+        {
+            _cancellationTokenSources.RemoveIfContainsKey(guid);
+        }
+
+        public async Task EnableDrainModeAsync(CancellationToken cancellationToken)
+        {
+            if (!IsDrainModeEnabled)
+            {
+                IsDrainModeEnabled = true;
+                _logger.LogInformation($"DrainMode is set to {IsDrainModeEnabled}");
+
+                CancelFunctionInvocations();
+
+                List<Task> tasks = new List<Task>();
+
+                _logger.LogInformation("Calling StopAsync on the registered listeners");
+                foreach (IListener listener in Listeners)
+                {
+                    tasks.Add(listener.StopAsync(cancellationToken));
+                }
+
+                await Task.WhenAll(tasks);
+
+                _logger.LogInformation("Call to StopAsync complete, registered listeners are now stopped");
+            }
+        }
+
+        public void CancelFunctionInvocations()
+        {
+            foreach (var keyValuePair in _cancellationTokenSources)
+            {
+                string invocationId = keyValuePair.Key.ToString();
+                try
+                {
+                    _logger?.LogInformation("Requesting cancellation for function invocation '{invocationId}'", invocationId);
+                    keyValuePair.Value.Cancel();
+                }
+                catch (ObjectDisposedException)
+                {
+                    _logger?.LogInformation("Cancellation token for function invocation '{invocationId}' already disposed. No action required", invocationId);
+                }
+                catch (Exception e)
+                {
+                    _logger?.LogInformation("Exception occured when attempting to request cancellation for function invocation '{invocationId}', '{e.GetType().Name}:{e.Message}'", invocationId, e.GetType().Name, e.Message);
```

Comment:
Why not log error and pass the exception here?

Reply:
pass the exception where, Do you mean throw the exception?

Reply:
Pass into the LogError call.